src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java

```diff
@@ -0,0 +1,782 @@
+                new AssemblyContigWithFineTunedAlignments.Serializer();
+        private static final BasicInfo.Serializer basicInfoSerializer = new BasicInfo.Serializer();
+        private static final Jump.Serializer jumpSerializer = new Jump.Serializer();
+        private static final ReferenceSegmentsAndEventDescription.Serializer eventsSerializer =
+                new ReferenceSegmentsAndEventDescription.Serializer();
+
+        final AssemblyContigWithFineTunedAlignments tigWithInsMappings;
+        final BasicInfo basicInfo;
+        private List<Jump> jumps;
+        private List<SimpleInterval> eventPrimaryChromosomeSegmentingLocations;
+        private ReferenceSegmentsAndEventDescription referenceSegmentsAndEventDescription;
+        private byte[] altSeq;
+
+        /**
+         * These getters may return null if the corresponding fields are not properly initialized yet.
+         */
+        List<Jump> getJumps() {
+            return jumps;
+        }
+        List<SimpleInterval> getEventPrimaryChromosomeSegmentingLocations() { return eventPrimaryChromosomeSegmentingLocations;}
+        ReferenceSegmentsAndEventDescription getSegmentsAndDescription() { return referenceSegmentsAndEventDescription; }
+        byte[] getAltSeq() {
+            return altSeq;
+        }
+
+        AnnotatedContig(final AssemblyContigWithFineTunedAlignments tigWithInsMappings, final SAMSequenceDictionary refSequenceDictionary) {
+
+            final AlignedContig sourceTig = tigWithInsMappings.contig;
+            final List<AlignmentInterval> deOverlappedAlignmentConfiguration =
+                    deOverlapAlignments(sourceTig.alignmentIntervals, refSequenceDictionary);
+            final AlignedContig contig = new AlignedContig(sourceTig.contigName, sourceTig.contigSequence, deOverlappedAlignmentConfiguration,
+                                            sourceTig.hasEquallyGoodAlnConfigurations);
+            this.tigWithInsMappings = new AssemblyContigWithFineTunedAlignments(contig, tigWithInsMappings.insertionMappings);
+
+            this.basicInfo = new BasicInfo(contig);
+
+            annotate(refSequenceDictionary);
+        }
+
+        private static List<AlignmentInterval> deOverlapAlignments(final List<AlignmentInterval> originalAlignments,
+                                                                   final SAMSequenceDictionary refSequenceDictionary) {
+            final List<AlignmentInterval> result = new ArrayList<>(originalAlignments.size());
+            final Iterator<AlignmentInterval> iterator = originalAlignments.iterator();
+            AlignmentInterval one = iterator.next();
+            while (iterator.hasNext()) {
+                final AlignmentInterval two = iterator.next();
+                // TODO: 11/5/17 an edge case is possible where the best configuration contains two alignments,
+                //       one of which contains a large gap, and since the gap split happens after the configuration scoring,
```

Comment:
I agree it is backwards. But...

The reason was that the (naive) alignment configuration scoring module rightnow uses MQ and AS (aligner score) for picking the "best" configuration (i.e. sub-list of the alignments given by aligner), which would be technically wrong if we were to split the gap and to simply grab the originating alignment's values.

This is especially true for AS, whose recomputing takes more time, and code, and forces us to know how AS are computed in the aligner so that there's no bias in computing the scores of naive alignments vs gap-split alignments (may not matter in practice, but still takes more code to compute).

Lots of the code in the discovery stage was devoted actually to alignment related acrobatics and edge cases so that the breakpoints we could resolve are as accurate as possible. 
I've kept in mind your wisdom that different aligners may be experimented with, but it seems unlikely in the near future (their own quirkiness, lack of API for JNI, etc); it seems more and more likely to me that eventually it's inevitable to have a custom alignment module in a high-quality SV pipeline, but again, the near future has other top priorities.

What do you think?

Reply:
added a line of doc briefly explaining the reason